components/eamxx/src/physics/cosp/eamxx_cosp.cpp

```diff
@@ -131,87 +138,101 @@ void Cosp::run_impl (const double dt)
 
   // Call COSP wrapper routines
   if (update_cosp) {
+    // Get fields from field manager; note that we get host views because this
+    // interface serves primarily as a wrapper to a c++ to f90 bridge for the COSP
+    // all then need to be copied to layoutLeft views to permute the indices for
+    // F90.
+
+    // Ensure host data of input fields is up to date
+    get_field_in("qv").sync_to_host();
+    get_field_in("qc").sync_to_host();
+    get_field_in("qi").sync_to_host();
+    get_field_in("sunlit").sync_to_host();
+    get_field_in("surf_radiative_T").sync_to_host();
+    get_field_in("T_mid").sync_to_host();
+    get_field_in("p_int").sync_to_host();
+    get_field_in("cldfrac_rad").sync_to_host();
+    get_field_in("eff_radius_qc").sync_to_host();
+    get_field_in("eff_radius_qi").sync_to_host();
+    get_field_in("dtau067").sync_to_host();
+    get_field_in("dtau105").sync_to_host();
+
+    // Compute z_mid
+    const auto T_mid_d = get_field_in("T_mid").get_view<const Real**>();
+    const auto qv_d  = get_field_in("qv").get_view<const Real**>();
+    const auto p_mid_d = get_field_in("p_mid").get_view<const Real**>();
+    const auto phis  = get_field_in("phis").get_view<const Real*>();
+    const auto pseudo_density = get_field_in("pseudo_density").get_view<const Real**>();
+    const auto z_mid_d = m_z_mid.get_view<Real**>();
+    const auto z_int = m_z_int.get_view<Real**>();
+    const auto ncol = m_num_cols;
+    const auto nlev = m_num_levs;
+
+    using KT       = KokkosTypes<DefaultDevice>;
+    using ExeSpace = typename KT::ExeSpace;
+    using ESU      = ekat::ExeSpaceUtils<ExeSpace>;
+    using PF       = scream::PhysicsFunctions<DefaultDevice>;
+
+    const auto scan_policy = ESU::get_thread_range_parallel_scan_team_policy(ncol, nlev);
+    const auto g = physics::Constants<Real>::gravit;
+    Kokkos::parallel_for(scan_policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+        const int i = team.league_rank();
+        const auto p_mid_s = ekat::subview(p_mid_d, i);
+        const auto T_mid_s = ekat::subview(T_mid_d, i);
+        const auto qv_s = ekat::subview(qv_d, i);
+        const auto z_int_s = ekat::subview(z_int, i);
+        const auto z_mid_s = ekat::subview(z_mid_d, i);
+        const Real z_surf  = phis(i) / g;
+        const auto pseudo_density_s = ekat::subview(pseudo_density, i);
+
+        // 1. Compute dz (recycle z_mid_s as a temporary)
+        const auto dz_s = z_mid_s; // 
+        PF::calculate_dz(team, pseudo_density_s, p_mid_s, T_mid_s, qv_s, dz_s);
+        team.team_barrier();
+
+        // 2. Compute z_int (vertical scan)
+        PF::calculate_z_int(team,nlev,dz_s,z_surf,z_int_s);
+        team.team_barrier();
+
+        // 3. Compute z_mid (int->mid interpolation)
+        PF::calculate_z_mid(team,nlev,z_int_s,z_mid_s);
+        team.team_barrier();
+    });
+
+    const auto T_mid_h = get_field_in("T_mid").get_view<const Real**, Host>();
```

Comment:
These are all host views, but only some of them get the `_h` suffix for host. Would it be better to rename all of them, or default to no suffix for host views since most of these are host?

Reply:
Yeah, I thought about it, then decided to leave as is. But you're right, to avoid confusion, since we have some host and some dev, it's best to be consistent. I will add `_h` to all host views.